### ISSUE TITLE
[FEA] Implementation of lfilter.

### ISSUE THREAD

**Issue by luigifreitas** [rapidsai/cusignal#7]
The `lfilter` function would be really useful for demodulation. For example, FM demodulation (de-emphasis and stereo).

Scipy Signal:
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.lfilter.html#scipy.signal.lfilter

**Comment by awthomp** [comment]
I concur with this feature request, and it's a high value item for the next version of cuSignal. @mnicely took an initial look at lfilter and can sum up some of his findings in the next few weeks.

That said, @luigifreitas -- could you provide a few examples in Scipy Signal that you'd like to use in cuSignal? Will you be using any initial conditions? 2D or just 1D? Etc.

**Comment by mnicely** [comment]
My initial attempt to the lfilter used in SciPy Signal to the GPU was unsuccessful due to data dependencies. I've attached my attempt at simplifying the math. I can give it another look at the beginning of the year.

A possible next approach might be to find an alternate filter that implemented in parallel, but the answers might vary slightly. 

[lfilter.pdf](https://github.com/rapidsai/cusignal/files/4005630/lfilter.pdf)

**Comment by awthomp** [comment]
@mnicely -- let's get a test case from @luigifreitas and work backwards. We can potentially get other folks to look at it as well. I haven't looked at the lfilter source in a bit, but I remember there being a significant number of options/algorithmic routes that we may be able to simplify/ignore in the short term. If we get a toy example working, maybe we can go from there? Either way, we'll revisit in January.

**Comment by luigifreitas** [comment]
I've created a [Gist](https://gist.github.com/luigifreitas/fe412f2640b79c9c6f9de8e52a9caa61) with a working notebook with all functions necessary to demodulate Stereo WBFM. As requested, this example is using Numpy and SciPy Signal components. The Pilot clock recovery is using my custom PLL based on Numpy. The implementation is working well but I'm still working on it. In short, this will require the following functions from SciPy Signal:

- [Hilbert Transform](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.hilbert.html)
- [Butterworth Filter Design](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.butter.html)
- [Linear Filter](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.signal.lfilter.html#scipy.signal.lfilter)

I already have a Manchester Decoder for RDS working but it still needs more polish. I'll update the Gist once it's ready.

**Comment by awthomp** [comment]
Thanks, @luigifreitas!

The Hilbert transform is included in cuSignal. You can find it [here](https://github.com/rapidsai/cusignal/blob/d2b8ddf157c1659be78e073534883031ffff2423/cusignal/signaltools.py#L1052)

I'll look at the Butterworth filter design and continue development on the linear filter. Cheers!

**Comment by awthomp** [comment]
Also, @luigifreitas - do you need initial conditions for lfilter? Or can we start with Zi=None?

**Comment by luigifreitas** [comment]
> Also, @luigifreitas - do you need initial conditions for lfilter? Or can we start with Zi=None?

I don't think we are going to need initial conditions for a narrow band signal like WBFM. Cheers!

**Comment by luigifreitas** [comment]
On second thought, the linear filter initial conditions are kinda nice to have. Not a direct requirement but should improve filtering.

**Comment by luigifreitas** [comment]
I found a [blog post](https://raphlinus.github.io/audio/2019/02/14/parallel-iir.html) written by @raphlinus regarding parallel evaluation of IIR filters. He suggests using a matrix approach instead of the serial direct form like Transposed DF-II used by Scipy's `lfilter`. I don't see small deviations being a problem.

**Comment by awthomp** [comment]
We are currently working on GPU accelerating the DF-II structure used in `lfilter`. As @mnicely pointed out, there's a lot of data dependence that may be difficult to accelerate, but I'd like to see the perf characterization of the standard approach before moving onto more 'researchy' techniques that may deviate from 'standard results'

I've completed a first pass as the raw Python code here: https://github.com/awthomp/cusignal-dev/blob/lfilter_dev/python/cusignal/_lfilter.py

Note: this is not necessarily optimized Python code and hasn't been extensively tested on larger arrays. We also need to add normalization.

@luigifreitas -- I'd appreciate a representative workflow with synthetic data of sizes that interest you.

**Comment by luigifreitas** [comment]
Sounds great, @awthomp! I wrote a [notebook](https://github.com/rapidsai/cusignal/files/4149176/lfilter.zip) using the lfilter with synthetic data.

**Comment by awthomp** [comment]
@mnicely and I have been actively working GPU acceleration of the DFII filter over the past few weeks. Unfortunately, there's a data dependency that makes the fundamental algorithm difficult to implement on GPU, specifically:

```python
y[n] -= ((a[nn - count] * y[n-k]) / a[0])
```

In this instance, the previous value of y[n], i.e. y[n-k] needs to be updated and accessible to the current 'batch of threads' doing the process. Since this can't be guaranteed in parallel programming, this is -- as we believe -- a fundamentally serial algorithm.

We'd love to understand alternatives to `lfilter`.

**Comment by awthomp** [comment]
Closing until we can identify alternatives the the data dependent DF-II algorithm.

**Comment by awthomp** [comment]
I was looking at PyTorch Audio, and it looks like they have a version of lfilter that runs on GPU: https://github.com/pytorch/audio/blob/master/torchaudio/functional.py#L673-L744

A few points:
* a and b must be the same size (e.g. pad with zeros if needed)
* input signal needs to be scaled between -1 and 1

Rough code:

```python
import torchaudio
import numpy as np
from scipy import signal
import torch

# CPU
sig = np.arange(10)/10
a = [1.0, 0.25, 0.5]
b = [1.0, 0, 0]
y = signal.lfilter(b, a, sig)
print(y)

# GPU (PyTorch)
torch.set_default_tensor_type('torch.cuda.DoubleTensor')
torch_sig = torch.as_tensor(sig)
torch_a = torch.as_tensor(a)
torch_b = torch.as_tensor(b)
torch_y = torchaudio.functional.lfilter(torch_sig, torch_a, torch_b)

print(torch_y)
print(torch_y.is_cuda)
```

Going to reopen this issue.

**Comment by rossant** [comment]
FYI we also have [an FFT-based GPU implementation of lfilter using cupy here](https://github.com/MouseLand/Kilosort2/blob/master/pykilosort/cptools.py#L179-L275)

**Comment by awthomp** [comment]
This is great, @rossant ! Do you have any perf/accuracy metrics between your FFT based lfilter and scipy.signal.lfilter?

What would you suggest for cuSignal? Looks like torchaudio is using the DF-II implementation, and I was going to muck something together to look at low-level perf.

BTW - We've added new 1D and 2D correlate/convolution functionality with raw CuPy CUDA kernels that deliver 2-3x over the previous GPU-perf.

**Comment by luigifreitas** [comment]
I made a quick comparison between the two methods. The `torchaudio` based `lfilter` takes significantly longer to complete. Both results seem to be practically equal.

![image](https://user-images.githubusercontent.com/6627901/80754280-26300500-8b05-11ea-85d8-a8300141428d.png)

```
%timeit gfilt = cs.lfilter(b, a, sig)
293 µs ± 1.35 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit tfilt = torchaudio.functional.lfilter(torch_sig, torch_a, torch_b)
178 ms ± 2.04 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

**Comment by awthomp** [comment]
Hey @luigifreitas -- Thanks for the input! Do you mind posting the code you used to generate this?

**Comment by luigifreitas** [comment]
Hi @awthomp, sure! I added four types of comparisons in this [Jupyter Notebook](https://github.com/rapidsai/cusignal/files/4561639/lfilter_torch.zip). I also ported the `lfilter` [function](https://pytorch.org/audio/_modules/torchaudio/functional.html#lfilter) from Torch to CuPy. The performance is worst but it can still be optimized.

```
# Scipy Signal
%time scipy_filt = cs.lfilter(b, a, sig)
CPU times: user 757 µs, sys: 234 µs, total: 991 µs
Wall time: 467 µs

# Torch Signal w/ CUDA
%time torch_cuda_filt = torchaudio.functional.lfilter(torch_cuda_sig, torch_cuda_a, torch_cuda_b)
CPU times: user 458 ms, sys: 1.2 ms, total: 459 ms
Wall time: 461 ms

# Torch Signal w/ CPU
%time torch_cpu_filt = torchaudio.functional.lfilter(torch_cpu_sig, torch_cpu_a, torch_cpu_b)
CPU times: user 223 ms, sys: 7.95 ms, total: 231 ms
Wall time: 187 ms

# Torch Function Converted to CuPy
%time cupy_filt = cupy_lfilter(cupy_sig, cupy_a, cupy_b)
CPU times: user 762 ms, sys: 20.1 ms, total: 782 ms
Wall time: 787 ms
```

```python
def cupy_lfilter(waveform, a_coeffs, b_coeffs):
    # pack batch
    shape = waveform.shape
    waveform = waveform.reshape(-1, shape[-1])

    assert (a_coeffs.shape[0] == b_coeffs.shape[0])
    assert (len(waveform.shape) == 2)

    dtype = waveform.dtype
    n_channel, n_sample = waveform.shape
    n_order = a_coeffs.shape[0]
    n_sample_padded = n_sample + n_order - 1
    assert (n_order > 0)

    # Pad the input and create output
    padded_waveform = cp.zeros((n_channel, n_sample_padded), dtype=dtype)
    padded_waveform[:, (n_order - 1):] = waveform
    padded_output_waveform = cp.zeros((n_channel, n_sample_padded), dtype=dtype)

    # Set up the coefficients matrix
    # Flip coefficients' order
    a_coeffs_flipped = cp.flip(a_coeffs, 0)
    b_coeffs_flipped = cp.flip(b_coeffs, 0)

    # calculate windowed_input_signal in parallel
    # create indices of original with shape (n_channel, n_order, n_sample)
    window_idxs = cp.expand_dims(cp.arange(n_sample), 0) + cp.expand_dims(cp.arange(n_order), 1)
    window_idxs = cp.tile(window_idxs, (n_channel, 1, 1))
    window_idxs += (cp.expand_dims(cp.expand_dims(cp.arange(n_channel), -1), -1) * n_sample_padded)
    #window_idxs = window_idxs.long()

    # (n_order, ) matmul (n_channel, n_order, n_sample) -> (n_channel, n_sample)
    input_signal_windows = cp.matmul(b_coeffs_flipped, cp.take(padded_waveform, window_idxs))

    for i_sample, o0 in enumerate(input_signal_windows[0,:]):
        windowed_output_signal = padded_output_waveform[:, i_sample:(i_sample + n_order)]
        o0 = cp.subtract(o0, windowed_output_signal.dot(a_coeffs_flipped))
        o0 = cp.divide(o0, a_coeffs[0])
        padded_output_waveform[:, i_sample + n_order - 1] = o0

    output = cp.clip(padded_output_waveform[:, (n_order - 1):], a_min=-1., a_max=1.)

    # unpack batch
    output = output.reshape(shape[:-1] + output.shape[-1:])

    return output
```

**Comment by awthomp** [comment]
Hey @luigifreitas -- great work! I think there's some room for optimization in the cupy'd version of Torch's `lfilter` too.

How about you write a PR referencing this issue and place your `cupy_lfilter` in `python/cusignal/filtering/filtering.py` as `lfilter`? You'll also need to edit the `__init__.py` file in filtering and also in `cusignal/python/cusignal`. 

When submitting the PR, be sure to label the title as [WIP] and update the CHANGELOG.md. 

We can obviously help on this stuff.

If you make the branch of your cuSignal fork writable, I can push to that same fork to update the PR as we look to optimize it.

We'll see if we can get the perf of this guy fast enough to merge!

**Comment by luigifreitas** [comment]
Hey @awthomp, just created the pull request as requested. Let me know if I missed something.

**Comment by Mayur28** [comment]
Hi,

I'm trying to use the torchaudio lfilter and I have been trying to compare the results to those produced by Librosa - Oddly, the results are quite different. If anyone could please offer any insight into this, it would be highly appreciated.

I have used the following code to experiment:

```
from scipy import signal
import librosa
import numpy as np
import torch
import torchaudio

pre_emphasis = -0.97
a_coeff = [1.0, 0.0]
b_coeff = [1.0, pre_emphasis]

#Librosa
audio1 = librosa.core.load('audio.wav', sr=16000)[0]
lib_ans = signal.lfilter(b_coeff, a_coeff, audio1)
print("Librosa - Min : {}     Max : {}".format(np.min(lib_ans), np.max(lib_ans)))

# Torchaudio
aud2 = torchaudio.load('audio.wav')[0].T # I have ensured that audio.wav has sampling_rate = 16000KHz using ffmpeg
torch_ans = torchaudio.functional.lfilter(aud2, torch.FloatTensor(a_coeff),torch.FloatTensor(b_coeff), clamp=False)
print("Torchaudio - Min : {}     Max : {}".format(torch.min(torch_ans), torch.max(torch_ans)))

```

Thanks!

**Comment by awthomp** [comment]
Hi @Mayur28 --

Thanks for posting a question. We deprecated the `lfilter` in cuSignal due to performance issues with real-time streaming data. As a work around, we support [FIR filtering](https://docs.rapids.ai/api/cusignal/stable/api.html?highlight=firfilt#module-cusignal.filtering.firfilter).

That said, I don't see cusignal used in your code. Perhaps file an issue to torchaudio and cross reference with scipy?

Hope this helps!
Adam

**Comment by Mayur28** [comment]
Hi @awthomp,

Apologies for the error. I will do that.
